File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Sections/AudioRoom/View/Chat/PanoChatView.m
//
//  PanoChatView.m
//  PanoAudioChatSDK
//
//  Created by pano on 2021/7/19.
//

#import "PanoChatView.h"
#import "PanoMessage.h"
#import "PanoChatTextCell.h"
#import "PanoClientService.h"

@interface PanoChatView () <UITableViewDataSource, UITableViewDelegate, PanoChatDelegate>

@end

@implementation PanoChatView {
    
    UITableView *tableView;

    NSMutableArray <PanoMessage *>* dataSource;

    NSMutableArray <PanoMessage *>* pendingMessages;

    void *specificKey;
    
    dispatch_queue_t receiveMessageQueue;
}


- (void)initViews {
    tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
    [tableView registerClass:[PanoChatTextCell class] forCellReuseIdentifier:@"cellID"];
    tableView.tableFooterView = [UIView new];
    tableView.tableHeaderView = [UIView new];
    tableView.dataSource = self;
    tableView.delegate = self;
    tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    tableView.backgroundColor = [UIColor clearColor];
    tableView.showsVerticalScrollIndicator = false;
    tableView.allowsSelection = false;
    [self addSubview: tableView];
    
}

- (void)initConstraints {
    [tableView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(0);
    }];
}

- (void)start {
    dataSource = [NSMutableArray array];
    pendingMessages = [NSMutableArray array];
    receiveMessageQueue = dispatch_queue_create("com.pvc.receiveMessageQueue", NULL);
    [PanoClientService.service.chatService start];
    PanoClientService.service.chatService.delegate = self;
}

#pragma mark -- PanoChatDelegate
- (void)onReceiveMessage:(PanoMessage *)msg {
    dispatch_async(receiveMessageQueue, ^{
        NSUInteger count = self->pendingMessages.count;
        [self->pendingMessages addObject:msg];
        if (count == 0) {
            [self processPendingMessages];
        }
    });
}

- (void)processPendingMessages {
    NSUInteger count = pendingMessages.count;
    if (count == 0) {
        return;
    }
    __block CGFloat width = 0.0;
    pano_main_sync_safe(^{
        if (self->tableView.isDecelerating || self->tableView.isDragging) {
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), self->receiveMessageQueue, ^{
                [self processPendingMessages];
            });
            return;
        }
        width = self->tableView.bounds.size.width - 50;
    });
    for (NSInteger index=0; index<pendingMessages.count; index++) {
        [pendingMessages[index] caculateWidth:width];
    }
    NSArray *result = [[NSArray alloc] initWithArray:pendingMessages copyItems:true];
    [pendingMessages removeAllObjects];
    dispatch_async(dispatch_get_main_queue(), ^{
        [self addMessages:result];
    });
}

- (void)addMessages:(NSArray<PanoMessage *> *)messages {
    NSUInteger count = dataSource.count;
    [dataSource addObjectsFromArray:messages];
    NSMutableArray *inserts = [NSMutableArray array];
    
    for (NSInteger i=count; i<count+messages.count; i++) {
        [inserts addObject:[NSIndexPath indexPathForRow:i inSection:0]];
    }
    [tableView beginUpdates];
    [tableView insertRowsAtIndexPaths:inserts withRowAnimation:UITableViewRowAnimationFade];
    [tableView endUpdates];
    [tableView layoutIfNeeded];
    [tableView scrollToRowAtIndexPath:[NSIndexPath indexPathForRow:dataSource.count - 1 inSection:0] atScrollPosition:UITableViewScrollPositionBottom animated:true];
}


#pragma mark -- UITableViewDataSource, UITableViewDelegate
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return dataSource.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    PanoChatTextCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cellID" forIndexPath:indexPath];
    PanoMessage *message = dataSource[indexPath.row];
    cell->contentLabel.text = message.content;
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    PanoMessage * message = dataSource[indexPath.row];
    CGFloat height = message.size.height + 20;
    return height;
}

@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Sections/AudioRoom/View/Chat/PanoChatView.swift
//
//  PanoChatView.swift
//  PanoAudioChat
//
//  Created by wenhua.zhang on 2020/11/4.
//

import UIKit

class PanoChatView: PanoBaseView {
    
    var tableView: UITableView!
    
    var dataSource = [PanoMessage]()
    
    var pendingMessages = [PanoMessage]()
    
    let specificKey = DispatchSpecificKey<String>()
    
    let receiveMessageQueue = DispatchQueue(label: "com.pvc.receiveMessageQueue")
    
    override func initViews() {
        
        tableView = UITableView()
        tableView.register(PanoChatTextCell.self, forCellReuseIdentifier: "cellID")
        tableView.tableFooterView = UIView()
        tableView.tableHeaderView = UIView()
        tableView.dataSource = self
        tableView.delegate = self
        tableView.separatorStyle = .none
        tableView.backgroundColor = UIColor.clear
        tableView.showsVerticalScrollIndicator = false
        tableView.allowsSelection = false
        self.addSubview(tableView)
        
    }
    
    override func initConstraints() {
        tableView.snp.makeConstraints { (make) in
            make.edges.equalToSuperview()
        }
    }
    
    func initService() {
        PanoChatService.service()?.start()
        PanoChatService.service()?.delegate = self
        receiveMessageQueue.setSpecific(key: specificKey, value: "receiveMessageQueue")
    }
}

extension PanoChatView: UITableViewDataSource,  UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        dataSource.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cellID", for: indexPath) as! PanoChatTextCell
        let message = dataSource[indexPath.row]
        cell.contentLabel.text = message.content
        return cell
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        let message = dataSource[indexPath.row]
        let height = message.size.height + 20
        print("cellHeight->", height)
        return height
    }
}


func pano_main_sync_safe(block: ()->Void) {
    if Thread.isMainThread {
        block()
    } else {
        DispatchQueue.main.sync {
            block()
        }
    }
}

extension PanoChatView: PanoChatDelegate{
    
    func onReceiveMessage(msg: PanoMessage) {
        receiveMessageQueue.async {
            
            let count = self.pendingMessages.count;
            self.pendingMessages.append(msg);
            if count == 0 {
                self.processPendingMessages()
            }
        }
    }
    
    func processPendingMessages() {
        let count = self.pendingMessages.count;
        if count == 0 {
            return
        }
        var width : CGFloat = 0.0
        pano_main_sync_safe {
//            print("self.tableView.isDecelerating->", self.tableView.isDecelerating, "self.tableView.isDragging->",self.tableView.isDragging )
            if (self.tableView.isDecelerating || self.tableView.isDragging) {
                receiveMessageQueue.asyncAfter(deadline: DispatchTime.now() + DispatchTimeInterval.seconds(1)) {
                    self.processPendingMessages()
                }
                return
            }
            width = self.tableView.bounds.size.width - 50
        }
        
        for index in 0..<self.pendingMessages.count {
            self.pendingMessages[index].caculateWidth(width: width)
        }
        let result = self.pendingMessages
        self.pendingMessages.removeAll()
        DispatchQueue.main.async {
            self.addMessages(messages: result)
        }
    }
    
    func addMessages(messages: [PanoMessage]) {
        let count = self.dataSource.count
        self.dataSource.append(contentsOf: messages)
        var inserts = [IndexPath]()
        for index in count..<count+messages.count {
            inserts.append(IndexPath(row: index, section: 0 ))
        }
        
        self.tableView.beginUpdates()
        self.tableView.insertRows(at: inserts, with: .fade)
        self.tableView.endUpdates()
        self.tableView.layoutIfNeeded()
        let indexPath = IndexPath(row: self.dataSource.count - 1 , section: 0 )
        self.tableView.scrollToRow(at: indexPath, at: .bottom, animated: true)
    }
}
